// @swimlane/ngx-datatable/esm2015/lib/components/row-detail/row-detail-template.directive.js
import { __decorate } from "tslib";
import { Directive, TemplateRef } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
let DatatableRowDetailTemplateDirective = class DatatableRowDetailTemplateDirective {
    constructor(template) {
        this.template = template;
    }
};
DatatableRowDetailTemplateDirective.ɵfac = function DatatableRowDetailTemplateDirective_Factory(t) { return new (t || DatatableRowDetailTemplateDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef)); };
DatatableRowDetailTemplateDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: DatatableRowDetailTemplateDirective, selectors: [["", "ngx-datatable-row-detail-template", ""]] });
DatatableRowDetailTemplateDirective.ctorParameters = () => [
    { type: TemplateRef }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DatatableRowDetailTemplateDirective, [{
        type: Directive,
        args: [{
                selector: '[ngx-datatable-row-detail-template]'
            }]
    }], function () { return [{ type: ɵngcc0.TemplateRef }]; }, null); })();
export { DatatableRowDetailTemplateDirective };
