// @swimlane/ngx-datatable/esm2015/lib/components/body/body-group-header-template.directive.js
import { __decorate } from "tslib";
import { Directive, TemplateRef } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
let DatatableGroupHeaderTemplateDirective = class DatatableGroupHeaderTemplateDirective {
    constructor(template) {
        this.template = template;
    }
};
DatatableGroupHeaderTemplateDirective.ɵfac = function DatatableGroupHeaderTemplateDirective_Factory(t) { return new (t || DatatableGroupHeaderTemplateDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef)); };
DatatableGroupHeaderTemplateDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: DatatableGroupHeaderTemplateDirective, selectors: [["", "ngx-datatable-group-header-template", ""]] });
DatatableGroupHeaderTemplateDirective.ctorParameters = () => [
    { type: TemplateRef }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DatatableGroupHeaderTemplateDirective, [{
        type: Directive,
        args: [{
                selector: '[ngx-datatable-group-header-template]'
            }]
    }], function () { return [{ type: ɵngcc0.TemplateRef }]; }, null); })();
export { DatatableGroupHeaderTemplateDirective };
